internal/transport/http2_client.go

```diff
@@ -743,6 +744,12 @@ func (t *http2Client) NewStream(ctx context.Context, callHdr *CallHdr) (_ *Strea
 		case <-t.goAway:
 			return nil, errStreamDrain
 		case <-t.ctx.Done():
+			t.mu.Lock()
+			err := t.closeErr
+			t.mu.Unlock()
+			if err != nil {
+				return nil, err
+			}
```

Comment:
I don't believe the lock is necessary, but it's harmless in any case.  `t.cancel()` is called only after `t.closeErr` is set, and `t.closeErr` can only be set once, before `t.cancel()`.

Reply:
`t.ctx` is derived from a parent context, so `t.cancel()` isn't the only way to cancel `t.ctx`. And the read and write may race in the other case.
(That parent context is `ClientConn.ctx`, so this only happens when the ClientConn is closed)

Reply:
How is this different?

And I don't like returning nil here. Even if it should never happen.